ruby/command-t/ext/command-t/scanner.c

```diff
@@ -0,0 +1,181 @@
+// Copyright 2010-present Greg Hurrell. All rights reserved.
+// Licensed under the terms of the BSD 2-clause license.
+
+#include <ruby.h>
+#include <assert.h>
+#include <errno.h>
+#include <stdlib.h>  /* for qsort() */
+#include <string.h>  /* for strncmp() */
+#include <sys/mman.h>
+#include "scanner.h"
+#include "match.h"
+#include "matcher.h"
+#include "heap.h"
+#include "ext.h"
+#include "ruby_compat.h"
+
+// order matters; we want this to be evaluated only after ruby.h
+#ifdef HAVE_PTHREAD_H
+#include <pthread.h> /* for pthread_create, pthread_join etc */
+#endif
+
+typedef struct {
+    VALUE source;
+    size_t bufsize;
+    matches_t matches;
+} paths_t;
+
+void mark_paths(paths_t *paths) {
+    rb_gc_mark(paths->source);
+}
+
+void free_paths(paths_t *paths) {
+    munmap(paths, paths->bufsize);
+}
+
+VALUE CommandTPaths_from_array(VALUE klass, VALUE source) {
+    Check_Type(source, T_ARRAY);
+    rb_obj_freeze(source);
+
+    long len = RARRAY_LEN(source);
+    long bufsize = sizeof(paths_t) + len * sizeof(match_t);
+    paths_t *paths = mmap(NULL, bufsize,
```

Comment:
> It relies on virtual memory to prevent reallocations and copies but I think that it will work on all modern OSes. That will need to be tested.

If we want to pre-check for the existence of `mmap` at build time we can add something to `extconf.rb` (it's going to be something like [`have_func`](http://ruby-doc.org/stdlib-2.0.0/libdoc/mkmf/rdoc/MakeMakefile.html#method-i-have_func) or similar). But yes, we should have some fallback in place if it turns out that this isn't going to work on one of the major platforms.